api/pkg/webhooks/manager.go

```diff
@@ -0,0 +1,205 @@
+package webhooks
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+)
+
+type WebhookManager interface {
+	InvokeWebhooks(
+		context.Context,
+		EventType,
+		interface{},
+		func([]byte) error,
+		func(error) error,
+	) error
+}
+
+type SimpleWebhookManager struct {
+	WebhookClients map[EventType][]WebhookClient
+}
+
+// InvokeWebhooks iterates through the webhooks for a given event and invokes them.
+// Sync webhooks are called first, and only after all of them succeed, the async webhooks are called.
+// Sync webhooks are called in the order that they are defined. The call order of async webhooks are
+// is not guaranteed.
+// If any of the sync clients are set to abort, the whole chain aborts as long as 1 sync request returns error.
+// onSuccess and onError are callbacks that are called after all webhooks are invoked.
+// For sync clients, the payload can be either the original input payload, or the response from another sync webhook.
+// This can be specified in the UseDataFrom field
+// For async clients, the payload is only the original input payload.
+// Only one webhook's response can be used as the finalResponse
+func (w *SimpleWebhookManager) InvokeWebhooks(
+	ctx context.Context,
+	event EventType,
+	p interface{},
+	onSuccess func([]byte) error,
+	onError func(error) error,
+) error {
+	var asyncClients []WebhookClient
+	var syncClients []WebhookClient
+	var finalResponse []byte
+	whc, ok := w.WebhookClients[event]
+	if !ok {
+		return fmt.Errorf("Could not find event %s", event)
+	}
+	originalPayload, err := json.Marshal(p)
+	if err != nil {
+		return err
+	}
+	for _, client := range whc {
+		if client.IsAsync() {
+			asyncClients = append(asyncClients, client)
+		} else {
+			syncClients = append(syncClients, client)
+		}
+	}
+
+	// Mapping to store response from different webhooks
+	responsePayloadLookup := make(map[string][]byte)
+
+	for _, client := range syncClients {
+		var tmpPayload []byte
+		if client.GetUseDataFrom() == "" {
+			tmpPayload = originalPayload
+		} else if tmpPayload, ok = responsePayloadLookup[client.GetUseDataFrom()]; !ok {
+			// This should only happen if a previous error had an error, but did not abort
+			// and the current client is trying to use the response from that client
+			return fmt.Errorf(
+				"webhook name %s not found, this could be because an error in a downstream webhook",
```

Comment:
Just curious, is 'upstream' like webhooks called before? Or the other way round? I can't get my head around this haha

Reply:
Fixed